File: Container With Most Water - Leetcode 11.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        n = len(height)
        l = 0
        r = n - 1
        max_area = 0

        while l < r:
            w = r - l
            h = min(height[l], height[r])
            a = w * h
            max_area = max(max_area, a)
            
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1

        return max_area

# Time Complexity: O(n)
# Space Complexity: O(1)

File: Container With Most Water - Leetcode 11/Container With Most Water - Leetcode 11.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int n = height.size();
        int l = 0;
        int r = n - 1;
        int maxArea = 0;

        while (l < r) {
            int w = r - l;
            int h = min(height[l], height[r]);
            int a = w * h;
            maxArea = max(maxArea, a);

            if (height[l] < height[r]) {
                l++;
            } else {
                r--;
            }
        }

        return maxArea;
    }
};
